selfdrive/loggerd/loggerd.cc

```diff
@@ -335,138 +286,66 @@ int main(int argc, char** argv) {
     SubSocket * sock = SubSocket::create(s.ctx, it.name);
     assert(sock != NULL);
     poller->registerSocket(sock);
-    socks.push_back(sock);
-
-    for (int cid=0; cid<=MAX_CAM_IDX; cid++) {
-      if (std::string(it.name) == cameras_logged[cid].frame_packet_name) {
-        s.rotate_state[cid].fpkt_sock = sock;
-      }
-    }
     qlog_states[sock] = {.counter = 0, .freq = it.decimation};
   }
 
   Params params;
 
   // init logger
   logger_init(&s.logger, "rlog", true);
+  logger_rotate();
   params.put("CurrentRoute", s.logger.route_name);
 
   // init encoders
-  pthread_mutex_init(&s.rotate_lock, NULL);
-
+  s.last_camera_seen_tms = millis_since_boot();
   // TODO: create these threads dynamically on frame packet presence
   std::vector<std::thread> encoder_threads;
   encoder_threads.push_back(std::thread(encoder_thread, LOG_CAMERA_ID_FCAMERA));
-  s.rotate_state[LOG_CAMERA_ID_FCAMERA].enabled = true;
+  if (cameras_logged[LOG_CAMERA_ID_FCAMERA].trigger_rotate) {
+    s.max_waiting += 1;
+  }
 
   if (!Hardware::PC() && params.getBool("RecordFront")) {
     encoder_threads.push_back(std::thread(encoder_thread, LOG_CAMERA_ID_DCAMERA));
-    s.rotate_state[LOG_CAMERA_ID_DCAMERA].enabled = true;
+    if (cameras_logged[LOG_CAMERA_ID_DCAMERA].trigger_rotate) {
+      s.max_waiting += 1;
+    }
   }
   if (Hardware::TICI()) {
     encoder_threads.push_back(std::thread(encoder_thread, LOG_CAMERA_ID_ECAMERA));
-    s.rotate_state[LOG_CAMERA_ID_ECAMERA].enabled = true;
+    if (cameras_logged[LOG_CAMERA_ID_ECAMERA].trigger_rotate) {
+      s.max_waiting += 1;
+    }
   }
 
-  uint64_t msg_count = 0;
-  uint64_t bytes_count = 0;
-  AlignedBuffer aligned_buf;
+  auto rotator = std::thread(rotator_thread);
 
-  double start_ts = seconds_since_boot();
-  double last_rotate_tms = millis_since_boot();
-  double last_camera_seen_tms = millis_since_boot();
+  uint64_t msg_count = 0, bytes_count = 0;
+  double start_ts = millis_since_boot();
   while (!do_exit) {
-    // TODO: fix msgs from the first poll getting dropped
     // poll for new messages on all sockets
     for (auto sock : poller->poll(1000)) {
-
       // drain socket
-      Message * last_msg = nullptr;
-      while (!do_exit) {
-        Message * msg = sock->receive(true);
-        if (!msg) {
-          break;
-        }
-        delete last_msg;
-        last_msg = msg;
-
-        QlogState& qs = qlog_states[sock];
-        logger_log(&s.logger, (uint8_t*)msg->getData(), msg->getSize(), qs.counter == 0 && qs.freq != -1);
-        if (qs.freq != -1) {
-          qs.counter = (qs.counter + 1) % qs.freq;
-        }
-
+      QlogState &qs = qlog_states[sock];
+      Message *msg = nullptr;
+      while (!do_exit && (msg = sock->receive(true))) {
+        const bool in_qlog = qs.freq != -1 && (qs.counter++ % qs.freq == 0);
+        logger_log(&s.logger, (uint8_t *)msg->getData(), msg->getSize(), in_qlog);
```

Comment:
logger_log will block logger_next wile writing. If we don’t use lh_log here, an additional rotation thread will not help much.

Reply:
But that's only while a single message is logged. That should be fine, right? I don't want to wait till a while socket is drained.

Reply:
how about check  `if (s.waiting_rotate == s.max_waiting)`  in while loop after logger_log ? It may be better than checking it every 1ms in another thread

Reply:
Guess I can put that all in a helper function and call that.